File: 0733-flood-fill/0733-flood-fill.cpp
class Solution {
public:
    void dfs(int i,int j,int n,int m,int color,vector<vector<int>>&image){
        int orgColor = image[i][j];
        image[i][j] = color;

        int del[4][2] = {{1,0},{-1,0},{0,-1},{0,1}};

        for(int k = 0; k < 4; k++){
            int row = i + del[k][0];
            int col = j + del[k][1];

            if(row < 0 || row >= n || col < 0 || col >= m)continue;

            if(image[row][col] == orgColor && orgColor != color){
			    dfs(row,col,n,m,color,image);
		    }
        }
    }

    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int n = image.size();
        int m = image[0].size();
        // dfs(sr,sc,n,m,color,image);
        
        queue<pair<int,int>>q;
        int orgColor = image[sr][sc];

        q.push({sr,sc});
        image[sr][sc] = color;

        while(!q.empty()){
            pair<int,int> p = q.front();
            q.pop();

            int i = p.first;
            int j = p.second;

            int del[4][2] = {{1,0},{-1,0},{0,-1},{0,1}};

            for(int k = 0; k < 4; k++){
                int row = i + del[k][0];
                int col = j + del[k][1];

                if(row < 0 || row >= n || col < 0 || col >= m)continue;

                if(image[row][col] == orgColor && orgColor != color){
                    image[row][col] = color;
                    q.push({row,col});
                }
            }

        }

        return image;
    }
};

File: 0733-flood-fill/0733-flood-fill.java
class Solution {
    void dfs(int sr,int sc,int[][] image,int orgColor,int color){
        image[sr][sc] = color;
        int[][] del = {{1,0},{-1,0},{0,1},{0,-1}};

        // if we get color == orgColor and there is no condition in floodfill function like this(if(color == image[sr][sc])return image;) then it will only run for 4 times no dfs call will made
        for(int k = 0; k < 4; k++){
            int nrow = sr + del[k][0];
            int ncol = sc + del[k][1];

            if(nrow < 0 || ncol < 0 || nrow == image.length || ncol == image[0].length || image[nrow][ncol] != orgColor || orgColor == color)continue;

            dfs(nrow,ncol,image,orgColor,color);
        }
    }
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        //  we can avoid the condition in the if of dfs function -> orgColor == color by putting if(color == orgColor) return image in this function
        // if(color == image[sr][sc])return image; // it will save only four iterations
        dfs(sr,sc,image,image[sr][sc],color);
        return image;
    }
}